juju/juju
# Initial support for IPv6

Adds support for IPv6, but without requiring IPv6.

There are some known issues that should not break, or be broken by, this
commit:

 * Manual is the only supportable provider for IPv6
 * RFC 6761 (6.3), expecting localhost to resolve to both the IPv4 and
   IPv6 loopback addresses, is not always satisfied in Ubuntu builds
   (lp:1644009), but
 * Ubuntu adds both IPv4 and IPv6 loopback addresses regardless of the
   version specified for lo in ENI, so loopback will resolve to a usable
   loopback address in any case
 * Code to be revised when these issues are resolved in Ubuntu are
   commented with `TODO(macgreagoir) IPv6. ...`
 * A bug around `enabled-ha` using existing machines (lp:1642618)
   requires a work-around for HA

QA steps:

 * IPv4, simply bootstrap as usual
 * IPv6
   - Build a machine with a working IPv6 address (which may require something like an IPv6 tunnel from Hurricane Electric) and no IPv4 addresses apart from loopback. (Feel welcome to contact me about this directly.)
   - Deploy to this machine with the manual provider
 * The expected result is a working environment in both IP version cases (noting the HA bug with the manual provider)
